### PR TITLE
CTCTRADERS-1224: SafetyAndSecurity XML + Submission Model

### PR DESCRIPTION
**Pull request by MatthewAlexanderBeck**

### CODE REVIEW

File: test/generators/JourneyModelGenerators.scala [hunk 1]
```diff
@@ -264,11 +338,29 @@ trait JourneyModelGenerators {
...
+
+      containers <- if (containersUsed) { nonEmptyListOf[Container](maxNumberOfItemsLength).map(Some(_)) } else Gen.const(None)
+
+      specialMentions <- Gen.option(nonEmptyListOf[SpecialMention](maxNumberOfItemsLength))
+
+      producedDocuments <- if (isDocumentTypeMandatory) { nonEmptyListOf[ProducedDocument](maxNumberOfItemsLength).map(Some(_)) } else Gen.const(None)
+    } yield ItemSection(itemDetail, itemConsignor, itemConsignee, packages, containers, specialMentions, producedDocuments)
+  }
+
+  def genItemSectionOld(containersUsed: Boolean = false, addDocument: Boolean = false, circumstanceIndicator: Option[String] = None): Gen[ItemSection] = {
```
Comment (PSouthern) [inline test/generators/JourneyModelGenerators.scala:363]: Should we have old and new sections in the model?  Or can we refactor replace the old with the new?

File: test/models/journeyDomain/SafetyAndSecuritySpec.scala [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+import models.{Index, UserAnswers}
+import org.scalacheck.Gen
+import org.scalatest.TryValues
+import pages.ModeAtBorderPage
+import pages.safetyAndSecurity._
+
+class SafetyAndSecuritySpec extends SpecBase with GeneratorSpec with TryValues with JourneyModelGenerators {
+
+  "SafetyAndSecurity can be parsed from UserAnswers" - {
+
```
Comment (PSouthern) [inline test/models/journeyDomain/SafetyAndSecuritySpec.scala:36]: Should we have a spec to cover this scenario?

```
Does specific circumstance 
indicator = 
(E) Authorised
Economic Operators?
```

File: app/models/journeyDomain/SafetyAndSecurity.scala [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+    ModeAtBorderPage.optionalReader.flatMap {
+      case Some("4") | Some("40") => ConveyanceReferenceNumberPage.reader.map(Some(_))
+      case _ =>
+        AddConveyanceReferenceNumberPage.reader.flatMap {
+          bool =>
+            if (bool) ConveyanceReferenceNumberPage.reader.map(Some(_)) else none[String].pure[UserAnswersReader]
+        }
+    }
+
+  private def placeOfUnloading: UserAnswersReader[Option[String]] =
```
Comment (PSouthern) [inline app/models/journeyDomain/SafetyAndSecurity.scala:94]: Should this include a condition to test the circumstance indicator value?

File: test/navigation/SpecialMentionsNavigatorSpec.scala [hunk 4]
```diff
@@ -231,13 +231,16 @@ class SpecialMentionsNavigatorSpec extends SpecBase with ScalaCheckPropertyCheck
...
+            .success
+            .value
 
           navigator
             .nextPage(AddAnotherSpecialMentionPage(index), NormalMode, userAnswers)
             .mustBe(controllers.addItems.routes.DocumentTypeController.onPageLoad(userAnswers.id, index, itemIndex, NormalMode))
         }
 
-        "to DocumentType when set to false and AddSecurityDetailsPage is 'Yes' and  AddCircumstanceIndicatorPage is 'Yes' and CircumstanceIndicator is either E, D, C and B" in {
+        "to DocumentType when set to false and AddSecurityDetailsPage is 'Yes' and  AddCircumstanceIndicatorPage is 'Yes' and CircumstanceIndicator is either E, D, C and B && AddCommercialReferenceNumberPage is false" in {
```
Comment (PSouthern) [inline test/navigation/SpecialMentionsNavigatorSpec.scala:243]: Can we add another spec to test this scenario please?

to DocumentType when set to false and AddSecurityDetailsPage is 'Yes' and  AddCircumstanceIndicatorPage is 'Yes' and AddCircumstanceIndicator is 'No' && AddCommercialReferenceNumberPage is false

**Review by PSouthern — COMMENTED** [review]
@MatthewAlexanderBeck @ashwinirudrappa Mostly questions added for clarification.  

Possibly a missing scenario in `Special Circumstances` navigator when testing the condition to navigate to Documents.

File: app/models/journeyDomain/SafetyAndSecurity.scala [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+import pages.ModeAtBorderPage
+import pages.safetyAndSecurity._
+
+case class SafetyAndSecurity(
+  circumstanceIndicator: Option[String],
+  paymentMethod: Option[String],
+  commercialReferenceNumber: Option[String],
+  conveyanceReferenceNumber: Option[String],
+  placeOfUnloading: Option[String],
+  consignor: Option[SecurityTraderDetails],
```
[the remark this answers is not in the document]
Comment (ashwinirudrappa) [reply app/models/journeyDomain/SafetyAndSecurity.scala:35]: @PSouthern itineraryList is a list of countryOfRouting

File: build.sbt [hunk 6]
```diff
@@ -87,7 +87,7 @@ lazy val root = (project in file("."))
   )
 
 lazy val testSettings: Seq[Def.Setting[_]] = Seq(
-  fork := true,
+  fork := false,
```
[the remark this answers is not in the document]
Comment (ashwinirudrappa) [reply build.sbt:90]: This is to fix that long java options errors that we get when we run our unit tests

File: app/navigation/SpecialMentionsNavigator.scala [hunk 7]
```diff
@@ -109,11 +109,15 @@ class SpecialMentionsNavigator @Inject()() extends Navigator {
...
-      case (Some(true), Some(true))  => userAnswers.get(CircumstanceIndicatorPage) map (CircumstanceIndicator.conditionalIndicators.contains(_))
-      case (Some(false), _)          => Some(false)
-      case _                         => None
+    (userAnswers.get(AddSecurityDetailsPage),
+     userAnswers.get(AddCircumstanceIndicatorPage),
+     userAnswers.get(AddCommercialReferenceNumberPage),
+     itemIndex.position == 0) match {
+      case (Some(true), Some(false), Some(false), true) => Some(true)
+      case (Some(true), Some(true), Some(false), true) =>
+        userAnswers.get(CircumstanceIndicatorPage) map (CircumstanceIndicator.conditionalIndicators.contains(_))
```
[the remark this answers is not in the document]
Comment (ashwinirudrappa) [reply app/navigation/SpecialMentionsNavigator.scala:118]: Yes it is covered in line number 116, it is that second Some(false)

File: test/generators/JourneyModelGenerators.scala (hunk 1 above)
Comment (ashwinirudrappa) [reply test/generators/JourneyModelGenerators.scala:363]: We will do it but later, not now

File: test/navigation/SpecialMentionsNavigatorSpec.scala (hunk 4 above)
Comment (ashwinirudrappa) [reply test/navigation/SpecialMentionsNavigatorSpec.scala:243]: Its covered in line 269 but not mentioned in test description there in the code

File: test/models/journeyDomain/SafetyAndSecuritySpec.scala (hunk 2 above)
Comment (MatthewAlexanderBeck) [reply test/models/journeyDomain/SafetyAndSecuritySpec.scala:36]: This is actually a really good spot, we havent added a check in the UserAnswers reader to look for this condition (though we do in the navigator). 

`  private def placeOfUnloading: UserAnswersReader[Option[String]] =
    AddPlaceOfUnloadingCodePage.reader.flatMap(
      bool => if (bool) PlaceOfUnloadingCodePage.optionalReader else none[String].pure[UserAnswersReader]
    )
`
There is the possibility here that the AddPlaceOfUnloadingCodePage might not be displayed if circumstance indicator is (E). This is a bug however due to time constraints I will probably need to add a sub task to address this later.

File: app/models/journeyDomain/SafetyAndSecurity.scala (hunk 3 above)
Comment (MatthewAlexanderBeck) [reply app/models/journeyDomain/SafetyAndSecurity.scala:94]: This is actually a really good spot. There is the possibility here that the AddPlaceOfUnloadingCodePage might not be displayed if circumstance indicator is (E). This is a bug however due to time constraints I will probably need to add a sub task to address this later

File: test/models/journeyDomain/SafetyAndSecuritySpec.scala (hunk 2 above)
Comment (MatthewAlexanderBeck) [reply test/models/journeyDomain/SafetyAndSecuritySpec.scala:36]: The below comment will add a test for this. Cheers.

File: test/navigation/SpecialMentionsNavigatorSpec.scala (hunk 4 above)
Comment (PSouthern) [reply test/navigation/SpecialMentionsNavigatorSpec.scala:243]: Resolved, covered in spec above.

**Review by PSouthern — APPROVED** [review]

**Review by amaalali — APPROVED** [review]

**Pull request merged**